### DISCUSSION TITLE
Is there a way to deref a quoted symbol?

### DISCUSSION THREAD

**Discussion by llmII** [janet-lang/janet#810] [General]
Trying to figure out a way to turn a function to a symbol (for the likes of `dyn`) and from a symbol to a function (for the ability to do `(type function)` and get `:function`).

If you have a function, that accepts a quoted symbol as an argument, where it then uses dyn to say change the doc on the function, yet only wants to do it if it's a function, it's at a loss for a way to convert the quoted symbol into the value at the unquoted symbol so as to determine the type.

So lets say I have a function, fun, and pass it to a function as an argument, arg. The function can't quote it, because it'll be quoted as "<function fun>". Same with stringing it. Same for symboling it. It always results in "<function fun>".

Same as above example, but instead of passing the function, I pass a quoted symbol of the function like "'fun" instead. There is no way to deref it into the actual function seemingly. If you unquote it, string it, symbol it, etc, and try to determine the resultant type with `type`, you just get `:symbol`.

**Comment by subsetpark** [comment]
I think there are a couple things worth establishing.

First: there's no need to set a symbol as a dynamic variable, instead of a function itself. Dynamic variables can be any datatype. So if you want to dynamically set a variable to point to a function, you can just do that directly.

Second: it might be useful to reorient your understanding of data types and (hopefully) simplify it a little bit. There's no such thing as a "quoted symbol" - there's just a symbol. Likewise, there's quoting, which can be applied to any data type. When you type:

```
foo
```

in the repl or in a script, the Janet interpreter parses that and the type of the resulting data is **symbol**. The interpreter/repl then automatically attempts to evaluate the data, and the evaluation of any symbol is some other value which the symbol is bound to in the environment.

Quoting, on the other hand, simply "defers" this evaluation step by one level. So if you type

```
'foo
```

It's parsed as `(quote foo)`, and then the same thing happens as in the first example. But in this case, there's a layer of `quote` that essentially "soaks up" the evaluation pass. When a quote is evaluated, the result is simply whatever's inside of it. And inside of it is the same symbol as before, `foo`. So now you have a single term in your program, but instead of some value in your environment, it's the value `foo`.

**Reply by subsetpark** [reply]
Finally, let me provide the most literal answer to your question: `eval` :) At any time you can call `eval` on any term, and that will follow the same evaluation semantics I was describing above. So if you have some symbol and you want to look up that symbol in the current environment and return the term it's bound to, you can simply call `(eval my-symbol)`.

**Reply by llmII** [reply]
> I think there are a couple things worth establishing.
> 
> First: there's no need to set a symbol as a dynamic variable, instead of a function itself. Dynamic variables can be any datatype. So if you want to dynamically set a variable to point to a function, you can just do that directly.

It's not that we're attempting to set a symbol as a dynamic variable, it's that `dyn` doesn't work on variables at all.
For instance:
```
(defn thing [] ())
(dyn thing)
```
gets a result of `nil`.

Changing it to:
```
(dyn 'thing)
```

Gets you: `@{:doc "(thing)\n\n" :source-map ("repl" 30 1) :value <function thing>}`

So say you have a function, that accepts as an argument a symbol, when you pass an unquoted symbol the argument becomes the value of the symbol passed as argument. You pass it a symbol created with `symbol` and that symbol is unbound to a value. The issue is, this function needs to get the value of the symbol bound to the argument, such that it can determine that the value referenced by that symbol, if it's a function, or not, and then take actions based upon "it is a function".

To illustrate:
```
(defn sym-fiddler
   [sym]
...
)

(defn thing [] ())

# sym-fiddler can't turn a function back into it's name, the symbol to
# which it was bound, so as to do (set ((dyn 'thing) :doc) "new documentation")
(sym-fiddler thing)
# sym-fiddler can't turn this back into the actual function to
# determine "hey, this is a function"
(sym-fiddler 'thing)
# symbol thing gets you <function thing>
(sym-fiddler (symbol thing)) 
# now you have an unbound symbol, thing
(sym-fiddler (symbol 'thing)) 
```

> Second: it might be useful to reorient your understanding of data types and (hopefully) simplify it a little bit. There's no such thing as a "quoted symbol" - there's just a symbol. Likewise, there's quoting, which can be applied to any data type. When you type:
> 
> ```
> foo
> ```
> 
> in the repl or in a script, the Janet interpreter parses that and the type of the resulting data is **symbol**. The interpreter/repl then automatically attempts to evaluate the data, and the evaluation of any symbol is some other value which the symbol is bound to in the environment.
> 
> Quoting, on the other hand, simply "defers" this evaluation step by one level. So if you type
> 
> ```
> 'foo
> ```
> 
> It's parsed as `(quote foo)`, and then the same thing happens as in the first example. But in this case, there's a layer of `quote` that essentially "soaks up" the evaluation pass. When a quote is evaluated, the result is simply whatever's inside of it. And inside of it is the same symbol as before, `foo`. So now you have a single term in your program, but instead of some value in your environment, it's the value `foo`.

I may be using the wrong way to express what I mean or may be of lack of understanding. I'm basically trying to make a distinction for the purposes of how `dyn` handles things in that it can take a `'sym` and get the metadata for `sym` but it can't take `sym` because it pulls in it's value rather than being able to look it up however it does. I do get the idea of quoting is to defer evaluation to a value.

Back to the hypothetical function I'm writing - I can pass it `'sym` and have `dyn` operate on it, but turning `'sym` into `sym` without `eval` is seemingly not possible. Which means I can't pass this function a `'sym` and then determine the value bound to the symbol once unquoted mostly because you can't unquote except where quoted or quasiquoted.

Fleshing out the above `sym-fiddler` (replacing the "..."):
```
# opt 1 - we got a symbol, needs eval... no other 
# way of doing it, replace ?? with eval (was hoping there 
# might be a better way than eval...)
(match (type (?? sym))
  :function  (set ((dyn sym) :doc) "new documentation"))
  _          (print "wasn't a function"))

# opt 2 - we got handed a value, no way to determine what symbol
# it was bound to without parsing the :value from the dyn
(match (type sym)
  # can't turn it into a symbol whose value is itself so that `dyn`
  # can operate on it
  :function true 
  _         (print "wasn't a function")
```

So as you said with "most literal", `eval`, which means probably double checking things about the value to make sure it's only a symbol, not a tuple or table or struct or so forth, to keep from possibly evaluating code, instead of just getting the value out of a symbol to determine the value's type.

**Reply by subsetpark** [reply]
Ok, I think I’m getting it. In that case - if you have some unknown argument, you can call symbol? And then eval and get whatever it’s currently bound to. But - by your own example, you can also use `dyn` to say “is this value currently bound to anything?” So you could also say (type (dyn my-argument)) and see if the result is :function. I’m not sure why you can’t take that approach here?

**Reply by llmII** [reply]
I was thinking along the lines of `(type (dyn my-argument))` but the issue is, `dyn` returns a table of metadata, not the value bound to the symbol it's asked about. 

And now that I look again, my issue was I was trying to test type on the table returned by `dyn` and didn't realize that the `:value` field in the table returned by `dyn` was the actual value vs a string or some other type... So it turns out you can get a value from a table returned by `dyn` and test against that, avoiding the whole `eval` rigamarole.

So basically I got hung up looking for "how to turn a quoted symbol" into a "symbol with a value bound to it" but it was just hidden away in "nesting" and after the 100th time reviewing the output of `(dyn 'sym)` I finally saw that the `:value` field is a value, not a human-ified representation of it, on which `type` may operate.

**Reply by subsetpark** [reply]
Ah yes, the guts of a `dyn` entry can be a little difficult at first! Glad you have something working.
